Wip/bl/adin2111 api

spectrum70 opened the pull request:
Hi,

i am extending the API support for this driver, as some of the added APIs are offered in the ADI no-os.
Not totally sure if some of the APIs are out of scope (not the Zephyr way to go), so collecting feedbacks on how to obtain same functionality where is the case. 

Thanks for  your support.

spectrum70 committed:
drivers: ethernet: eth_adin2111: adjust multicast filter

Multicast rule as 01:00:00:00:00:00 is actually shadowing
broadcast rule, enabling broadcast packets to be received/forwarded
even if broadcast rule/filter is not set.

Set the multicast filter with an appropriate mask.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: ethernet: eth_adin2111: extend api support

Extend api support with new functionalities.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: ethernet: eth_adin2111: enable promiscuous mode

Add promiscuous mode support.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

pdgendt commented:
Hey @spectrum70 there is a `set_config` driver API function you can use with `ETHERNET_CONFIG_TYPE_FILTER` to apply MAC filters.
See #69385 on recent changes to other drivers.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -27,6 +27,8 @@ LOG_MODULE_REGISTER(eth_adin2111, CONFIG_ETHERNET_LOG_LEVEL);
 
 #define DT_DRV_COMPAT adi_adin2111
 
+#define ADIN2111_ETH_MAC_ADDR_LEN		6U
```
Maybe it is time to have a MAC address length in the `include/zephyr/net/ethernet.h`?
@jukkar, @rlubos - what do you think?

jukkar replied:
Good point. I thought we already had a length defined there but I remembered wrong, please send a patch that defines one.

spectrum70 replied:
thanks, i can do it too

spectrum70 replied:
done

GeorgeCGV commented on drivers/ethernet/phy/phy_adin2111.c:
```diff
@@ -455,6 +455,28 @@ static int phy_adin2111_init(const struct device *dev)
 	return phy_adin2111_sft_pd(dev, false);
 }
 
+int phy_adin2111_mdio_read_c22(const struct device *dev, int reg_addr, uint16_t *val)
+{
+	return phy_adin2111_c22_read(dev, reg_addr, val);
+}
+
+int phy_adin2111_mdio_write_c22(const struct device *dev, int reg_addr, uint16_t val)
+{
+	return phy_adin2111_c22_write(dev, reg_addr, val);
+}
+
+int phy_adin2111_mdio_read_c45(const struct device *dev, uint16_t devad,
+			       uint16_t reg, uint16_t *val)
+{
+	return phy_adin2111_c45_read(dev, devad, reg, val);
+}
+
+int phy_adin2111_mdio_write_c45(const struct device *dev, uint16_t devad,
+				uint16_t reg, uint16_t val)
+{
+	return phy_adin2111_c45_write(dev, devad, reg, val);
+}
```
What is the goal here? There is [adin mdio driver](https://github.com/zephyrproject-rtos/zephyr/blob/be881d4cf269d3126722ce42e3c198b5960a3f3b/drivers/mdio/mdio_adin2111.c).

spectrum70 replied:
@GeorgeCGV  I see mdio driver works fine, but due to the big phy register set, idea is to have such c22/c45 functions as APIs too, and only way i found to have them accessible is by providing a *phy reference from adin2111 API and accessing them as phy APIs. Maybe the name should be phy_adin2111_write_c45() (without mdio).

GeorgeCGV replied:
You can try to get correctio MDIO like `mdio_shell.c` does:
`const struct device *mdio = DEVICE_DT_GET(DT_COMAT_GET_ANY_STATUS_OKAY(adi_adin2111_mdio));`

Then, using the MDIO API RW for both clauses is possible.

spectrum70 replied:
Thanks a lot, used mdio_read() and realted APIs.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -48,11 +50,93 @@ LOG_MODULE_REGISTER(eth_adin2111, CONFIG_ETHERNET_LOG_LEVEL);
+	return eth_adin2111_reg_write(dev, reg, val);
+}
+
+bool eth_adin2111_link_state(const struct device *dev, int port)
```
It would be better to provide a getter for the port's iface. That removes a need to know the iface index or name.

Then, the application can do what it needs: query link status, configure iface, etc...

spectrum70 replied:
It's ok. Can go that way.

spectrum70 replied:
done

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -935,13 +1019,14 @@ static int adin2111_filter_multicast(const struct device *dev)
 	const struct adin2111_config *cfg = dev->config;
 	bool is_adin2111 = (cfg->id == ADIN2111_MAC);
-	uint8_t mm[6] = {BIT(0), 0U,  0U, 0U, 0U, 0U};
+	uint8_t mm[6] = {BIT(0), 0U, 0x5EU, 0xFFU, 0xFFU, 0xFFU};
+	uint8_t mmask[6] = {0xFFU, 0xFFU, 0xFFU, 0x00U, 0x00U, 0x00U};
```
That prevents reception of LLDP, PTP, and others from the [01, 80, C2] (IEEE 802 group) required by some applications.

If there is a desire to limit multicasts to [01, 00, 5E], then it must be configurable.

spectrum70 replied:
@GeorgeCGV oh ok. Non really clear to me, may be wrong, but from testing, multicast 0x01 0x00 0x00 0x00 0x00 0x00 allows broadcast packets to be received too, so broadcast filter comes useless.

GeorgeCGV replied:
Indeed, currently used mask should be corrected to `0xFF, 0x0, 0x0, 0x0, 0x0, 0x0` instead of `0x1, 0x0, 0x0, 0x0, 0x0, 0x0`.
That should fix broadcast and multicast filters.

However, we don't want to limit the "kind" of the multicast. Therefore, the mac stays as `0x1, 0x0, 0x0, 0x0, 0x0, 0x0`.

spectrum70 replied:
ok, done, thanks

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -48,11 +50,93 @@ LOG_MODULE_REGISTER(eth_adin2111, CONFIG_ETHERNET_LOG_LEVEL);
+	if (ret) {
+		return ret;
+	}
+	ret = eth_adin2111_reg_write(dev, ADIN2111_SOFT_RST_REG, ADIN2111_SWRESET_KEY2);
```
The CS must be asserted for a minimum of 100ns. I am not sure how such a guarantee can be provided here.

spectrum70 replied:
ok, can modify this

spectrum70 replied:
applied same delay used in _init(), should be ok

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -48,11 +50,93 @@ LOG_MODULE_REGISTER(eth_adin2111, CONFIG_ETHERNET_LOG_LEVEL);
+	return eth_adin2111_reg_write(dev, ADIN2111_RESET, ADIN2111_RESET_SWRESET);
+}
+
+int eth_adin2111_mac_reset(const struct device *dev)
+{
```
> Note that PHY 1 must be out of software power-down for the MAC
subsystem reset to take effect

It is worth mentioning in the function comment.

spectrum70 commented:
> Hey @spectrum70 there is a `set_config` driver API function you can use with `ETHERNET_CONFIG_TYPE_FILTER` to apply MAC filters. See #69385 on recent changes to other drivers.

Thanks, changing as suggested.

spectrum70 committed:
net: ethernet: add definition for mac address length

Add a define for the mac address length, to be used as needed
in place of hardcoded 6.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

pdgendt commented:
> > Hey @spectrum70 there is a `set_config` driver API function you can use with `ETHERNET_CONFIG_TYPE_FILTER` to apply MAC filters. See #69385 on recent changes to other drivers.
> 
> Thanks, changing as suggested.

The mentioned PR has been merged in `main`.

spectrum70 commented:
Pushed a v2:
- fixed multicast filter/mask
- removed all APIs that are not really needed, mdio included
- used new mac hw filtering (ETHERNET_CONFIG_TYPE_FILTER)
- retested all by a simple unit test, and finally dhcpv4

thanks @GeorgeCGV and all.

jukkar commented on include/zephyr/net/ethernet.h:
```diff
@@ -49,10 +49,12 @@ extern "C" {
  * @{
```
Please change the subject of the mac address length commit to
`net: ethernet: add definition for mac address length`
so that it is easier to search for network related commits if needed.

spectrum70 replied:
Sorry, forgot to adjust it at the end, done now, thanks.

spectrum70 commented:
Pushed V.3
- fixed mac address length commit title prefix

GeorgeCGV commented:
The functions:

* `int eth_adin2111_reg_update(const struct device *dev, const uint16_t reg, uint32_t mask, uint32_t data);`
* `int eth_adin2111_sw_reset(const struct device *dev);`
* `int eth_adin2111_mac_reset(const struct device *dev);`
* `int eth_adin2111_broadcast_filter(const struct device *dev, bool enable)`

are exposed to the public, but they don't use the lock. That can lead to problems.
Also, it is not mentioned in the function documentation.

For the `port_idx`: we should stick with one name and type across the code base. My bad with `adin2111_read_fifo` on that.

spectrum70 commented:
> The functions:
> 
> * `int eth_adin2111_reg_update(const struct device *dev, const uint16_t reg, uint32_t mask, uint32_t data);`
> * `int eth_adin2111_sw_reset(const struct device *dev);`
> * `int eth_adin2111_mac_reset(const struct device *dev);`
> * `int eth_adin2111_broadcast_filter(const struct device *dev, bool enable)`
> 
> are exposed to the public, but they don't use the lock. That can lead to problems. Also, it is not mentioned in the function documentation.
> 
> For the `port_idx`: we should stick with one name and type across the code base. My bad with `adin2111_read_fifo` on that.

Thansk !  About the exposing without lock, there is a note about, so should be ok.
 * @note The caller is responsible for device lock.
 *       Shall not be called from ISR.

Fixed the port to port_idx, adding a patch to uniform this

spectrum70 committed:
drivers: ethernet: eth_adin2111: fix port to port_idx

Fix all references of port to port_idx to be uniform.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 commented:
Pushed V.4:
- fix port to port_idx reference for APIs
- fix port to port_idx references in the whole driver

GeorgeCGV commented:
Thank you, but certain points require attention.

#### Major

> The log is not in the correct place. When a filter exists, then the position is returned, otherwise `-ENOENT` or another negative error code.
> We shouldn't fail if the filter is already present. Therefore, returning > 0.

Related to `eth_adin2111_set_mac_filter`.

> About the exposing without lock, there is a note about, so should be ok.

The note is absent for:

> `int eth_adin2111_sw_reset(const struct device *dev);`
> `int eth_adin2111_mac_reset(const struct device *dev);`
> `int eth_adin2111_broadcast_filter(const struct device *dev, bool enable);`

#### Minor

* The `NET_ETH_ADDR_LEN` should be used, related to `adin2111_filter_multicast`.
* In `adin2111_port_set_config`: no need to convert promisc_mode to an int.
* The `port` naming is addressed, but not the type.
* Some code doesn't adhere to the existing codebase, such as implicit conditionals (`if (ret)`) and missing braces for single statement conditionals. Please refer to the PR comments.

spectrum70 committed:
drivers: ethernet: eth_adin2111: use NET_ETH_ADDR_LEN

Use new NET_ETH_ADDR_LEN definition.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 commented:
> Thank you, but certain points require attention.
> 
Thanks for the feedbacks, sorry.

> #### Major
> > The log is not in the correct place. When a filter exists, then the position is returned, otherwise `-ENOENT` or another negative error code.
> > We shouldn't fail if the filter is already present. Therefore, returning > 0.
> 
> Related to `eth_adin2111_set_mac_filter`.
> 

Thanks, fixed.

> > About the exposing without lock, there is a note about, so should be ok.
> 
> The note is absent for:
> 
> > `int eth_adin2111_sw_reset(const struct device *dev);`
> > `int eth_adin2111_mac_reset(const struct device *dev);`
> > `int eth_adin2111_broadcast_filter(const struct device *dev, bool enable);`
> 

Thanks, done.

> #### Minor
> * The `NET_ETH_ADDR_LEN` should be used, related to `adin2111_filter_multicast`.
done
> * In `adin2111_port_set_config`: no need to convert promisc_mode to an int.
done
> * The `port` naming is addressed, but not the type.
set all to const uint16_t as in the port_data structure.

> * Some code doesn't adhere to the existing codebase, such as implicit conditionals (`if (ret)`) and missing braces for single statement conditionals. Please refer to the PR comments.
Uhm, strange that it's not catched by checkpatch, i always pass checkpatch with --strict.

Fixed.

spectrum70 commented:
Pushed V5:
- fix eth_adin2111_set_mac_filter return value
- set all references to port_idx and const uint16_t everywhere
- add notes for locking in exported apis
- fix code styling issues
- use NET_ETH_ADDR_LEN always
- use promisc_mode

GeorgeCGV commented:
@spectrum70 LGTM, a few nitpicks left:

The comment of `ADIN2111_SW_RESET_DELAY_MS` can be simplified:

```c
/* As per RM rev. A page 20: approximately 10 ms (maximum) for internal logic to be ready.
 * Testing of Open Alliance mode shows 20 ms to be a safer value.
 */
```

However, I didn't experience any issues with a 10 ms delay. Could it be HW or application-related (_assuming the tests were done outside of the driver's init)_?

#### `eth_adin2111_find_filter`

Instead of `0xffffU` the `UINT16_MAX` definition can be used.

#### `eth_adin2111_set_mac_filter`

The line `1100` contains the final implicit condition.

spectrum70 commented:
Hi @GeorgeCGV ,

thanks,

> @spectrum70 LGTM, a few nitpicks left:
> 
> The comment of `ADIN2111_SW_RESET_DELAY_MS` can be simplified:
> 
> ```c
> /* As per RM rev. A page 20: approximately 10 ms (maximum) for internal logic to be ready.
>  * Testing of Open Alliance mode shows 20 ms to be a safer value.
>  */
> ```
> 
> However, I didn't experience any issues with a 10 ms delay. Could it be HW or application-related (_assuming the tests were done outside of the driver's init)_?

Comment fixed, at init() stage 10ms are ok, but i had to increment the value since in case of reset after configuration (in Open Alliance), i get errors until 15ms abt.

> 
> #### `eth_adin2111_find_filter`
> Instead of `0xffffU` the `UINT16_MAX` definition can be used.
> 

done

> #### `eth_adin2111_set_mac_filter`
> The line `1100` contains the final implicit condition.

agh, thanks :) done

spectrum70 commented:
Pushed V6:
- fix sw reset delay comment
- fix 0xffff to UINT16_MAX
- fix if (ret)

GeorgeCGV reviewed (dismissed):
Thank you for your contribution. Everything has been addressed. Perhaps the types of `int i, offset, reg, ret;` could be improved, but it's a minor issue.

spectrum70 commented:
@GeorgeCGV thanks a lot for all the help, i am still quite new to Zephyr, will improve :)

GeorgeCGV commented:
> Comment fixed, at init() stage 10ms are ok, but i had to increment the value since in case of reset after configuration (in Open Alliance), i get errors until 15ms abt.

Perhaps it would be better to have a delay argument in the sw reset function. This way, the initialization process could use the recommended RM 10ms delay, while users could provide a custom delay based on their use cases.

Some applications might have a requirement to boot as fast as possible. Therefore, unnecessary delays in `init` should be avoided.

spectrum70 commented:
> > Comment fixed, at init() stage 10ms are ok, but i had to increment the value since in case of reset after configuration (in Open Alliance), i get errors until 15ms abt.
> 
> Perhaps it would be better to have a delay argument in the sw reset function. This way, the initialization process could use the recommended RM 10ms delay, while users could provide a custom delay based on their use cases.
> 
> Some applications might have a requirement to boot as fast as possible. Therefore, unnecessary delays in `init` should be avoided.

Sounds good. Doing it.

spectrum70 commented:
Pushed V7:
- add a paramenter for custom delay in sw_reset api

GeorgeCGV reviewed (approved):
@spectrum70 that is nice, thank you!

spectrum70 commented:
Thanks all  !
